x-exception/src/X/Control/Monad/Catch.hs

```diff
@@ -43,23 +44,25 @@ bracketF a f g =
 --
 -- Exception and `Left` safe version of bracketEitherT.
 --
-bracketEitherT' :: MonadMask m => EitherT e m a -> (a -> EitherT e m c) -> (a -> EitherT e m b) -> EitherT e m b
+bracketEitherT'
+  :: forall m e a b c. MonadMask m
+  => EitherT e m a
+  -> (a -> EitherT e m c)
+  -> (a -> EitherT e m b)
+  -> EitherT e m b
 bracketEitherT' acquire release run =
-  EitherT $ bracketF
-    (runEitherT acquire)
-    (\r -> case r of
-      Left _ ->
-        -- Acquire failed, we have nothing to release
-        return . Right $ ()
-      Right r' ->
-        -- Acquire succeeded, we need to try and release
-        runEitherT (release r') >>= \x -> return $ case x of
-          Left err -> Left (Left err)
-          Right _ -> Right ())
-    (\r -> case r of
-      Left err ->
-        -- Acquire failed, we have nothing to run
-        return . Left $ err
-      Right r' ->
-        -- Acquire succeeded, we can do some work
-        runEitherT (run r'))
+  let
+    acquire' :: m (Either e a)
+    acquire' = runEitherT acquire
+
+    release' :: Either e a -> m (Either e c)
+    release' = either (return . Left) (runEitherT . release) -- if acquire' failed then pass through
+
+    run' :: Either e a -> m (Either e b) 
+    run' = either (return . Left) (runEitherT . run) -- if acquire' failed then skip run.
+
+ in EitherT $ mask $ \unmask -> do
+  resource <- acquire'
+  result <- unmask (run' resource) `onException` release' resource
+  releaseResult <- release' resource
```

Comment:
I dont think they're different? release always gets the output from the acquire thats how the type goes, and doesnt get access to the actual exception that might get thrown from the run.  I think the way the final errors are resolved in the last line matches teh bracket semantics.

In any case, its evidently not as simple or straight forward as i thought, so im happy to just close this PR